### DISCUSSION TITLE
Redirect to alternative frontends

### DISCUSSION THREAD

**Discussion by mispp** [uBlockOrigin/uBlock-issues#2666] [Ideas]
Would it be possible to have an option to redirect e.g. twitter.com to nitter.nl (different domain, same path)? Basically same as [privacy redirect](https://addons.mozilla.org/en-US/firefox/addon/privacy-redirect/) extension. The biggest impact would probably be in loading speed on mobile especially with medium mode enabled.

Not sure if this is a welcomed idea for an ad blocker, but given the goal in ublock origin readme it just might make sense:
> uBO's primary goal is to help users neutralize these privacy-invading methods in a way that welcomes those users who do not wish to use more technical means.

Edit: changed twitter.nl to nitter.nl

**Comment by stephenhawk8054** [comment]
Redirecting to new connection is a security-risk, so currently uBO does not support that.

**Comment by iam-py-test** [comment]
I personally think this should be left to a dedicated extension. Beyond the technical effort required to make the actual extension, maintaining the rules of what should redirect to what would probably require a lot of work. [There are also so many options](https://duckduckgo.com/?q=twitter+front+ends); how is the maintainer to decide which users would prefer. They could of course do nothing by default and have users pick, but then what's the point.
> Basically same as [privacy redirect](https://addons.mozilla.org/en-US/firefox/addon/privacy-redirect/) extension

Then why not use that?

**Reply by mispp** [reply]
It cannot be used on mobile.

**Reply by MasterKia** [reply]
Use Fennec from F-Droid and install that add-on as a custom add-on collection.

**Reply by krystian3w** [reply]
Better use updated: https://github.com/libredirect/browser_extension as long 2 years gap disappear from "<code class="notranslate">SimonBrazell/&shy;privacy-redirect</code>".

**Reply by mispp** [reply]
@krystian3w yeah, this is better than the one i linked

@MasterKia this needs an account as an extension developer if i remember correctly.

**Reply by krystian3w** [reply]
Yes, the collection requires an free Moz://a account* (same like to FirefoxSync) or a friend who will make you such a collection and add your suggestions (currently the collection mechanism spoils the installation of allowed add-ons as we omit them from the collection - that is, uninstalled uBlock Origin can throw an error from the web page despite the fact that since version 85 it is possible to install about 12 add-ons this way. I haven't checked since which release this has been happening and to what extent Fennec F-droid can mitigate it. It could also be a server-level bug, in which case any release with such a workaround has a problem).

\* - Only Chrome/Edge/Apple stories need paid account to Publish addons.

**Reply by mispp** [reply]
mozilla seems to not handle idea of mobile extensions well.. not sure what else to say here.
is there any sane reason to limit the list? also, fennec is not necessery, but i think also firefox beta would work.

if mozilla would allow more extension, this discussion wouldnt be relevant. libredirect fully supports initial idea, but since ublock origin is already approved to be installed, redirect could be easily implemented (i think)

**Reply by krystian3w** [reply]
There is no problem to use Firefox Beta with a specific collection of what does not exceed the limit of about 180 add-ons, as you do not need to reinstall about 12 allowed and each of the 11 items would be valuable.

The reason, supposedly, is to carefully implement the API and test how the add-on will slow down the mobile browser lest someone be left in the lurch that Firefox no longer opens on the phone and has no way to remove the buggy add-on.

**Comment by partingscientist** [comment]
Why not create and [use](https://github.com/gorhill/uBlock/wiki/Advanced-settings#userresourceslocation) your own scriptlet for it and host it on your own GitHub account? uBlock Origin is very flexible for these types of problems.

A lazy improvement on [this](https://github.com/uBlockOrigin/uBlock-issues/discussions/2387#discussion-4603923) would look like...

```javascript
/// redirect-frontend.js
/// alias rf.js
/**
 * @example
 * twitter.com##+js(rf, https://nitter.nl)
 * www.reddit.com##+js(rf, https://old.reddit.com)
 * */
(function() {
    "use strict";
    const host = "{{1}}";
    window.location.replace(host + window.location.pathname);
})();
```

You can customise it to your own liking and use case!

**Reply by krystian3w** [reply]
Perhaps his goal is to make it work from the next update on a click or immediately for some 50 million uBo users.

In addition, maybe it is layered too low and expects a bad portal to have zero page views, because everything loads with the help of the API and the scriptlet does not provide this (the main document loads and in the background some one resource/request according to Logger).

**Reply by partingscientist** [reply]
I mean, considering `stephenhawk8054` already answered their question (not supported), I'm just suggesting something they can personally use. Besides, their main goal as they posted seemed to be achieving better performance for mobile, with which the scriptlet can help by redirecting to a lighter frontend.

**Reply by mispp** [reply]
Suggestion for scriptlet sounds good, didn't know that it's possible. Will try it.
 
Regarding the answer that it is not supported, Statement is that it *currently* is not supported. It doesn't mean that it cannot be or will not be. The security concern is valid, but I don't know about this enough to have an opinion. 

One of the goals is better performance, but avoiding Twitter, YT and friends would increase our privacy since the providers still can track a lot of things just by visiting the page. This is why I quoted the statement from the official ublock origin readme. 

Additional benefit is much faster load on mobile. On desktop I can already use libredirect.

**Reply by stephenhawk8054** [reply]
The security risk is a bad actor among a list's maintainers can redirect any link to malicious links and the list will be auto updated without any notices to users.

And remember that even some instances could be unsafe someday either: https://github.com/libredirect/browser_extension/pull/43

**Reply by krystian3w** [reply]
Apparently, there is hope that MV3 will allow installation of any prescribed extension.

But recommended is again forking: https://github.com/libredirect/browser_extension/issues/510

**Comment by krystian3w** [comment]
uBo 1.66+[^1] with enabled "I trust this filters":
```adb
/^https?:\/\/www\.reddit\.com\/(?!media)/$doc,uritransform=/www/old/,to=www.reddit.com
|https://twitter.com^$doc,uritransform=/twitter/twiiit/
|https://www.twitter.com^$doc,uritransform=/(www\.)?twitter/twiiit/
|https://x.com^$doc,uritransform=/x/twiiit/
|https://www.x.com^$doc,uritransform=/(www\.)?x/twiiit/
! Polish:
||forum.dobreprogramy.pl^$doc,uritransform=/.*/https:\/\/dobreprogramy.freeforums.net\//
||frazpc.pl^|$doc,uritransform=/.*/https:\/\/www.boardpc.pl\//
||gery.pl^|$doc,uritransform=/.*/https:\/\/www.boardpc.pl\//
```

I do not tested with Chromium, works fine with Firefox (or other Gecko/Servo).

    On the downside, someone needs to add a mechanism that allows logging in (You probably need to switch to many `|https://...` or 2 regexes `/^https:\/\/...`).
    I haven't tested whether the old Reddit or regional Reddit with the old design has already disabled logging in (I think they could have done it so that it wouldn't be possible anymore).
    Certainly some engines don't like embedded frames (when someone forces you to open in a new tab), e.g. `nitter.space` (so with a dynamic domain you would have to do a separate removal of `/embed` to reduce the 404)

[^1]: It can only be withdrawn at any time, with forcing migration to uBOL if it remains there for longer.

**Reply by Dum4G** [reply]
You rock! Easy as that + enabling the checkbox
||x.com^$doc,uritransform=/x.com/nitter.net/
||twitter.com^$doc,uritransform=/twitter.com/nitter.net/

**Reply by Rudxain** [reply]
For reddit, I recommend this instead:
```adblock
||www.reddit.com/r/$doc,uritransform=/www/old/
```
because `i.redd.it/*` images redirect to `www.reddit.com/media/*`, which then redirs to Old, but Old can't handle the URL-params

**Reply by krystian3w** [reply]
Until they abandon `/r/` in 2027-2060 with reject 301 redirects.
